### ISSUE TITLE
If a podman VM is created via `libkrun`, all `podman machine` commands stop working

### ISSUE THREAD

**Issue by tumido** [containers/podman-desktop#9860]
Issue rewritten as per @cbr7 's request. Original bug report below.

### Bug description

Since https://github.com/podman-desktop/podman-desktop/pull/8247 we can now select `libkrun` as "Provider type" when creating a new Podman VM. This breaks everything using `podman machine *` commands **including Podman Desktop extensions** like Red Hat Authenticator etc...

If an extension uses a `podman machine` command under the hood, it is broken, since Podman Desktop doesn't propagate the `CONTAINERS_MACHINE_PROVIDER` environment variable it used when creating the VM.

### Steps to reproduce

Scenario "never leave Podman Desktop UI":

1. Use Apple silicon Mac
2. Use Podman Desktop to create a new podman machine VM
3. Select `libkrun` when creating the machine
4. Wait for it to start
5. I click a button in Podman Desktop to subscribe it to Red Hat subscription (via the Red Hat Authenticator plugin)
6. See in Tasks that it fails with error 125.

https://github.com/user-attachments/assets/1aa4091f-0688-4f39-bd1e-39d2e1c5d154

OR scenario "Follow Podman Desktop documentation":

We're following these 2 docs:
- https://podman-desktop.io/docs/podman/gpu (select MacOS tab)
- https://podman-desktop.io/docs/podman/adding-certificates-to-a-podman-machine

1. Use Apple silicon Mac
2. Use Podman Desktop to create a new podman machine VM
3. Select `libkrun` when creating the machine
4. Wait for it to start
5. Try `podman machine ssh` command
6. See error:

```
$ podman machine ssh
Error: vm podman-machine-default not found: podman-machine-default: VM does not exist
```

### Additional context

A simple workaround is to either propagate `CONTAINERS_MACHINE_PROVIDER` to `.profile|.zshrc|.bashrc` or create/update `~/.config/containers/containers.conf` with:

```
[machine]
  provider = "libkrun"
```

Ideal solution would be either guide user to do so when creating the Podman VM in Podman Desktop or do it for them.

Original Bug report below:
---

### Bug description

Since https://github.com/podman-desktop/podman-desktop/pull/8247 we can now select `libkrun` as "Provider type" when creating a new Podman VM. Selecting this option results in a properly working VM and podman with a single caveat - the `podman machine *` commands do not work out of the box anymore while podman itself works just fine:

```
$ podman machine list
NAME        VM TYPE     CREATED     LAST UP     CPUS        MEMORY      DISK SIZE

$ podman machine ssh
Error: vm podman-machine-default not found: podman-machine-default: VM does not exist

$ podman run -it registry.access.redhat.com/ubi9/toolbox cat /etc/redhat-release
Red Hat Enterprise Linux release 9.4 (Plow)
```

The way how the PR mentioned above implemented the `libkrun` option is that it uses an environment variable `CONTAINERS_MACHINE_PROVIDER=libkrun`.

This environment variable is never propagated anywhere else besides the machine install workflow and the user is also not made aware they need to apply this change for their future CLI **AND** Podman Desktop use.

This missing env variable has unintended sideffects for Podman Desktop itself, or rather its extensions - like Red Hat Authentication plugin never working, since it relies on `podman machine` commands [behind the scenes](https://github.com/redhat-developer/podman-desktop-redhat-account-ext/blob/b7b04c8e76bc406e7a38accd21442cc9ba5d68f6/src/podman-cli.ts#L26) to subscribe the machine.

### Operating system

macOS M1

### Installation Method

Installer from website/GitHub releases

### Version

1.14.0

### Steps to reproduce

1. Use Apple silicon Mac
2. Use Podman Desktop to create a new podman machine VM
3. Select `libkrun` when creating the machine
4. Wait for it to start
5. Try running `podman machine ssh` - it fails
6. Try running `podman machine list` - returns empty list

### Relevant log output

```shell

```

### Additional context

A simple workaround is to either propagate `CONTAINERS_MACHINE_PROVIDER` to `.profile|.zshrc|.bashrc` or create/update `~/.config/containers/containers.conf` with:

```
[machine]
  provider = "libkrun"
```

**Comment by cbr7** [comment]
@tumido This is a podman issue, not related to Podman Desktop, podman only displays machines for the current provider type, which is default `applehv` on macos, meaning that without changing the provider type value via the configuration file or environment variable, as you did in the workaround section, if you have a `libkrun` machine it will be invisible in podman cli. This has been reported and a requested change in podman for some time now, you can find the issue [here](https://github.com/containers/podman/issues/20702).

**Comment by tumido** [comment]
Hi @cbr7, I honestly disagree. I **didn't** "use" `podman` manually to init a new machine or export/changed the env variable myself. Instead I've used Podman Desktop to do that. I've "just" selected an option in Podman Desktop dropdown here:

![Image](https://github.com/user-attachments/assets/00ea21c5-d171-4de6-b41a-989dc87ce1ca)

It was Podman Desktop which chose to set it as an env variable when creating the machine. It didn't tell me or warn me that "things" will stop working. You see, there's nothing special about the dropdown either. As a user how am I supposed to know I have to change something manually? And that such env variable even exists? I'm not told that. If at least Podman Desktop would offer some kind of documentation or alert the user that they have to do manual changes in CLI... no, it does nothing like that.

And since then, the machine is invisible **not just to** podman in CLI but also transitively Podman Desktop and its extensions.

Honestly I don't see the following flow as "no issue" or "somebody else problem":

1. I create a machine in Podman Desktop
2. I click a button in Podman Desktop to subscribe it to Red Hat subscription (via the Red Hat Authenticator plugin)
3. It fails with error 125.

https://github.com/user-attachments/assets/1aa4091f-0688-4f39-bd1e-39d2e1c5d154


All of this is standard Podman Desktop user flow, I never left the Podman Desktop user story, not a single step to any side. So please don't dismiss this as "somebody else problem". It's Podman Desktop's doing.

The purpose of Podman Desktop is that user doesn't need to know that such env variable even exists. How do you expect them to know if Podman Desktop doesn't even tell them "things won't work unless you do this or that manually"?

**Comment by cbr7** [comment]
@tumido your video is not viewable so I can't really see what you are are saying.

In your initial report you paste the logs from some podman cli commands, I explained to you why podman cli does not see your `libkrun` machine and why when `podman machine list` command shows no machines.

Nowhere in your repro steps from OP did you do anything more than create the libkrun machine in PD, which yes is a featured offered from PD, but afterwards in steps 5 and 6 you try to run podman cli commands. If there is an issue with how Podman Desktop operates feel free to clarify what it is, but it's not obvious from OP.

EDIT: As I mentioned previously, if you want to interact with `libkrun` podman machine in podman cli you need to use the solution you have in the workaround section.

**Comment by tumido** [comment]
Shoot.. No idea what's wrong with the video.. I'm able to download it back and it plays on both my machines and phone.. Interesting. Whatever the reason is, I've tried reuploading it here (give Google some time to process it in higher resolution)

https://drive.google.com/file/d/1VVorAau3mAxP_oOYSinAk_12z2gtj3E_/view?usp=sharing

In my original report it reads:

> This missing env variable has unintended sideffects for Podman Desktop itself, or rather its extensions - like Red Hat Authentication plugin never working, since it relies on podman machine commands [behind the scenes](https://github.com/redhat-developer/podman-desktop-redhat-account-ext/blob/b7b04c8e76bc406e7a38accd21442cc9ba5d68f6/src/podman-cli.ts#L26) to subscribe the machine.

And that's why i went back to CLI to demonstrate.

Additionally, Podman Desktop documentation many times refers to `podman` CLI and `podman machine` commands when troubleshooting. I have never seen the claim that if I use Podman Desktop, I should **not** expect Podman CLI to work.

> As I mentioned previously, if you want to interact with libkrun podman machine in podman cli you need to use the solution you have in the workaround section.

While you're saying that, Podman Desktop never said that. Neither in the UI or in the documentation. And as you can see in the video, it breaks the UI flows as well, not just CLI.

**Comment by cbr7** [comment]
>Additionally, Podman Desktop documentation many times refers to podman CLI and podman machine commands when troubleshooting. I have never seen the claim that if I use Podman Desktop, I should not expect Podman CLI to work.

Podman Desktop helps manage podman machines, but that does not mean that one cannot work without the other, they are different projects with their own bugs and backlogs.

>While you're saying that, Podman Desktop never said that. Neither in the UI or in the documentation. And as you can see in the video, it breaks the UI flows as well, not just CLI.

I can see that something does indeed break, I would suggest you rewrite the issue in OP to clarify what the problem that you are reporting is, because it's not obvious from the initial report.

**Comment by tumido** [comment]
> Podman Desktop helps manage podman machines, but that does not mean that one cannot work without the other, they are different projects with their own bugs and backlogs.

That is understandable. However Podman Desktop should not intentionally fall apart and break thing for the user. Especially if there are known issues in `podman` itself that, it seems, the Podman Desktop team is aware of as you've mentioned earlier. If Podman Desktop allows users to select an option which has known sideeffects like: extensions not working, unable to follow other Podman Desktop guides, etc... it should at least be documented somewhere that this thing will happen.

> I can see that something does indeed break, I would suggest you rewrite the issue in OP to clarify what the problem that you are reporting is, because it's not obvious from the initial report.

Is this better now? I've added additional scenario, to demonstrate that the claim:

> Nowhere in your repro steps from OP did you do anything more than create the libkrun machine in PD, which yes is a featured offered from PD, but afterwards in steps 5 and 6 you try to run podman cli commands. If there is an issue with how Podman Desktop operates feel free to clarify what it is, but it's not obvious from OP.

..is actually a scenario supported by Podman Desktop documentation. Podman Desktop documentation asks me to run `podman machine ssh`. And it does not tell me that it is expected it won't work.

**Comment by cbr7** [comment]
>That is understandable. However Podman Desktop should not intentionally fall apart and break thing for the user. Especially if there are known issues in podman itself that, it seems, the Podman Desktop team is aware of as you've mentioned earlier. If Podman Desktop allows users to select an option which has known sideeffects like: extensions not working, unable to follow other Podman Desktop guides, etc... it should at least be documented somewhere that this thing will happen.

Are you sure the error is related to the libkrun podman machine at all? I've tried to repro the bug on my M2 Pro and the red hat extension works.

>..is actually a scenario supported by Podman Desktop documentation. Podman Desktop documentation asks me to run podman machine ssh. And it does not tell me that it is expected it won't work.

The documentation assumes podman is configured correctly, which in the case of using a libkrun podman machine implies that either the environment variable is used in the terminal you are running the commands or the configuration file has the correct provider type value.

@shipsing maybe some clarification in the docs for macos regarding `libkrun` is appropriate to prevent this confusion going forward given that podman cli is defaulting to `applehv` provider type and making `libkrun` podman machines seem invisible from CLI perspective without proper configuration.

**Comment by tumido** [comment]
> Are you sure the error is related to the libkrun podman machine at all? I've tried to repro the bug on my M2 Pro and the red hat extension works.

Yup pretty certain, setting `CONTAINERS_MACHINE_PROVIDER` solves the issue for me. Maybe you already have that set on on your machine? Or the config change applied? Undoing the change on my side makes the RH extension start failing again on Podman Desktop restart.

**Comment by cbr7** [comment]
>Don't you already have that set on on your machine?

No, I do not, the podman machine is invisible when checking `podman machine ls` in terminal.

**Comment by tumido** [comment]
Interesting.. are you on Podman Desktop v1.14.1 as well as I am? Is there anything I can provide to help you troubleshoot?


```
$ unset CONTAINERS_MACHINE_PROVIDER

$ cat ~/.config/containers/containers.conf
cat: /Users/tcoufal/.config/containers/containers.conf: No such file or directory

$ podman machine list
NAME        VM TYPE     CREATED     LAST UP     CPUS        MEMORY      DISK SIZE

$
```

![Image](https://github.com/user-attachments/assets/e7a3bf7a-7577-40c2-822f-96869c71ec15)

**Comment by cbr7** [comment]
@tumido no need, I was able to reproduce it on a different machine in the end, the authentication does fail when using `libkrun` with podman that is not correctly configured.

@jeffmaury seems that RH Auth extension cannot perform the authentication correctly when the user has a podman machine with a nondefault provider type if podman itself is not configured correctly for that provider, it only works after configuring podman to work correctly.

**Comment by benoitf** [comment]
For sso extension, it should be handled by https://github.com/redhat-developer/podman-desktop-redhat-account-ext/issues/44 and https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/182

**Comment by slemeur** [comment]
@dgolovin Please, can you confirm that the two issues on extension are fixing this issue?
